src/LendController.sol

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.13;
+
+import {ERC721, ERC721TokenReceiver} from 'solmate/tokens/ERC721.sol';
+import {SafeTransferLib, ERC20} from "solmate/utils/SafeTransferLib.sol";
+
+import {INFTLoanFacilitator} from './interfaces/INFTLoanFacilitator.sol';
+
+struct LoanTerms {
+    bool allowBuyouts;
+    bool allowLoanAmountIncrease;
+    uint16 interestRate;
+    uint128 amount;
+    uint32 durationSeconds;
+    address loanAsset;
+}
+
+interface NFTLoanTermsSource {
+    function terms(uint256 tokenId, address nft, address loanAsset, bytes calldata data) external returns (LoanTerms memory);
+}
+
+contract LendController is ERC721TokenReceiver {
+    using SafeTransferLib for ERC20;
+    
+    struct LenderSpec {
+        address lender;
+        uint256 amount;
+        bool pull;
+    }
+
+    struct BorrowRequest {
+        LenderSpec[] lendersSpecs;
+        address termsSource;
+        LoanTerms terms;
+    }
+
+    uint256 ONE = 1e18;
+    mapping(address => mapping(address => bool)) public termsApprovals;
+    mapping(address => mapping(address => uint256)) public lenderBalance;
+    mapping(uint256 => mapping(address => uint256)) public lenderLoanShares;
+
+    INFTLoanFacilitator public loanFacilitator;
+
+    constructor(INFTLoanFacilitator _facilitator) {
+        loanFacilitator = _facilitator;
+    }
+
+    event SetTermsApproval(address indexed from, address indexed termsContract, bool approved);
+
+    function setTermsApproval(address termsContract, bool approved) external {
+        termsApprovals[msg.sender][termsContract] = approved;
+
+        emit SetTermsApproval(msg.sender, termsContract, approved);
+    }
+
+    function purchaseNFT() external {
+        // allows anyone to call in and seize a seize NFT that
+        // the controller holds the lend ticket for and pays the lenders
+        // purchase price must be >= totalOwed on loan
+    }
+
+    function liquidateNFT(uint256 loanId) external {
+        // TBD if controller will have one liquidation mechanism 
+        // or will call out to the terms contract for implement
+        // update lenderBalance based on lenderLoanShares * sale value 
+    }
+
+    // used to lend to an exisiting loan
+    function lend(
+        uint256 loanId,
+        uint256 tokenId,
+        address nft,
+        BorrowRequest memory request,
+        bytes calldata data
+    ) external {
+        _lend(loanId, tokenId, nft, request, false, data);
+    }
+
+    function _lend(
+        uint256 loanId,
+        uint256 tokenId,
+        address nft,
+        BorrowRequest memory request,
+        bool skipIsBuyoutCheck,
+        bytes calldata data
+    ) internal {
+        LoanTerms memory terms = NFTLoanTermsSource(request.termsSource).terms(tokenId, msg.sender, request.terms.loanAsset, '');
```

Comment:
should this method make some checks as to whether `terms` from terms source match up with request.terms? ideally they're the same yes, but someone could maliciously pass mismatching?

Reply:
will get checked in NFTLoanFacilitator, right? Must meet or beat.